### ISSUE TITLE
camellia: padding incompatibilities with OpenSSL for CFB, OFB, and CTR modes

### ISSUE THREAD

**Issue by nmantani** [binref/refinery#24]
### Description
camellia command has padding incompatibilities with OpenSSL for CFB, OFB, and CTR modes. OpenSSL does not add padding and camellia command adds padding for these modes.

When a plaintext is encrypted with openssl command then the ciphertext output is decrypted with camellia command, the plaintext output is truncated. On the other hand, when a plaintext is encrypted with camellia command then the ciphertext output is decrypted with openssl command, the plaintext output has redundant trailing bytes. These behaviors do not occur for ECB and CBC modes. Truncation of plaintext outputs on decryption is an especially critical issue.

According to [the article of Block cipher mode of operation in the WikiPedia](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Padding), CFB, OFB and CTR modes do not require any special measures to handle messages whose lengths are not multiples of the block size. So I think it is incorrect that camellia command requires input as a multiple of the 16-byte block size for CFB, OFB and CTR modes.

I use Binary Refinery as Python modules for writing some plugins of [FileInsight-plugins](https://github.com/nmantani/FileInsight-plugins). Binary Refinery is so great to support a lot of encodings, crypto algorithms, and compression algorithms.

### To Reproduce
Encryption with ECB mode (identical output):
```sh
$ echo "This is a secret message." | openssl enc -e -camellia-128-ecb -K 30313233343536373839616263646566 | xxd
00000000: c4eb d417 c468 8165 2367 99f1 c5e0 bcb8  .....h.e#g......
00000010: ff47 867e 39ca b441 a4c3 0679 4a48 00bb  .G.~9..A...yJH..
$
```

```sh
$ echo "This is a secret message." | camellia --mode ecb -R 0123456789abcdef | xxd
00000000: c4eb d417 c468 8165 2367 99f1 c5e0 bcb8  .....h.e#g......
00000010: ff47 867e 39ca b441 a4c3 0679 4a48 00bb  .G.~9..A...yJH..
$
```
---

Encryption with CBC mode (identical output):
```sh
$ echo "This is a secret message." | openssl enc -e -camellia-128-cbc -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | xxd
00000000: cda7 a782 bce6 4ca8 6c3a a804 b141 ce6c  ......L.l:...A.l
00000010: c1c4 8c07 fb65 357d 21d0 4da9 6a2c 0f49  .....e5}!.M.j,.I
$
```

```sh
$ echo "This is a secret message." | camellia --iv 0123456789abcdef --mode cbc -R 0123456789abcdef | xxd
00000000: cda7 a782 bce6 4ca8 6c3a a804 b141 ce6c  ......L.l:...A.l
00000010: c1c4 8c07 fb65 357d 21d0 4da9 6a2c 0f49  .....e5}!.M.j,.I
$
```
---

Encryption with CFB mode (different outputs):
```sh
$ echo "This is a secret message." | openssl enc -e -camellia-128-cfb -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | xxd
00000000: 3e1a c0f8 aa74 e546 8925 eb0a 1776 1127  >....t.F.%...v.'
00000010: 6833 fe93 bb59 043d 8b3e                 h3...Y.=.>
$
```

```sh
$ echo "This is a secret message." | camellia --iv 0123456789abcdef --mode cfb -S 128 -R 0123456789abcdef | xxd
00000000: 3e1a c0f8 aa74 e546 8925 eb0a 1776 1127  >....t.F.%...v.'
00000010: 6833 fe93 bb59 043d 8b3e abce 0701 8677  h3...Y.=.>.....w
```
---

Encryption with OFB mode (different outputs):
```sh
$ echo "This is a secret message." | openssl enc -e -camellia-128-ofb -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | xxd
00000000: 3e1a c0f8 aa74 e546 8925 eb0a 1776 1127  >....t.F.%...v.'
00000010: 36dd fea5 869c 171f 547c                 6.......T|
```

```sh
$ echo "This is a secret message." | camellia --iv 0123456789abcdef --mode ofb -R 0123456789abcdef | xxd
00000000: 3e1a c0f8 aa74 e546 8925 eb0a 1776 1127  >....t.F.%...v.'
00000010: 36dd fea5 869c 171f 547c 0363 e05a 2850  6.......T|.c.Z(P
$
```
---
Encryption with CTR mode (different outputs):
```sh
$ echo "This is a secret message." | openssl enc -e -camellia-128-ctr -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | xxd
00000000: 3e1a c0f8 aa74 e546 8925 eb0a 1776 1127  >....t.F.%...v.'
00000010: 472b 60a4 8c98 0540 ce1a                 G+`....@..
$
```

```sh
$ echo "This is a secret message." | camellia --iv 0123456789abcdef --mode ctr -R 0123456789abcdef | xxd
00000000: 3e1a c0f8 aa74 e546 8925 eb0a 1776 1127  >....t.F.%...v.'
00000010: 472b 60a4 8c98 0540 ce1a dd81 0101 dc8f  G+`....@........
$
```
---
Encryption with openssl command and decryption with camellia command (ECB mode, works correctly):
```sh
$ echo "This is a secret message." | openssl enc -e -camellia-128-ecb -K 30313233343536373839616263646566 | camellia --mode ecb 0123456789abcdef | xxd
00000000: 5468 6973 2069 7320 6120 7365 6372 6574  This is a secret
00000010: 206d 6573 7361 6765 2e0a                  message..
$
```

Encryption with openssl command and decryption with camellia command (CBC mode, works correctly):
```sh
$ echo "This is a secret message." | openssl enc -e -camellia-128-cbc -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | camellia --iv 0123456789abcdef --mode cbc -S 128 -L 0123456789abcdef | xxd
00000000: 5468 6973 2069 7320 6120 7365 6372 6574  This is a secret
00000010: 206d 6573 7361 6765 2e0a                  message..
$
```

Encryption with openssl command and decryption with camellia command (CFB mode, the plaintext output is truncated):
```sh
$ echo "This is a secret message." | openssl enc -e -camellia-128-cfb -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | camellia --iv 0123456789abcdef --mode cfb -S 128 -L 0123456789abcdef | xxd
(22:20:55) warning in camellia: removing 10 bytes from the input to make it a multiple of the 16-byte block size
00000000: 5468 6973 2069 7320 6120 7365 6372 6574  This is a secret
$
```

Encryption with openssl command and decryption with camellia command (OFB mode, the plaintext output is truncated):
```sh
$ echo "This is a secret message." | openssl enc -e -camellia-128-ofb -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | camellia --iv 0123456789abcdef --mode ofb -L 0123456789abcdef | xxd
(22:23:24) warning in camellia: removing 10 bytes from the input to make it a multiple of the 16-byte block size
00000000: 5468 6973 2069 7320 6120 7365 6372 6574  This is a secret
$
```

Encryption with openssl command and decryption with camellia command (CTR mode, the plaintext output is truncated):
```sh
$ echo "This is a secret message." | openssl enc -e -camellia-128-ctr -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | camellia --iv 0123456789abcdef --mode ctr -L 0123456789abcdef | xxd
(22:24:28) warning in camellia: removing 10 bytes from the input to make it a multiple of the 16-byte block size
00000000: 5468 6973 2069 7320 6120 7365 6372 6574  This is a secret
$
```
---
Encryption with camellia command and decryption with openssl command (ECB mode, works correctly):
```sh
$ echo "This is a secret message." | camellia --mode ecb -R 0123456789abcdef | openssl enc -d -camellia-128-ecb -K 30313233343536373839616263646566 | xxd
00000000: 5468 6973 2069 7320 6120 7365 6372 6574  This is a secret
00000010: 206d 6573 7361 6765 2e0a                  message..
$
```

Encryption with camellia command and decryption with openssl command (CBC mode, works correctly):
```sh
$ echo "This is a secret message." | camellia --iv 0123456789abcdef --mode cbc -R 0123456789abcdef | openssl enc -d -camellia-128-cbc -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | xxd
00000000: 5468 6973 2069 7320 6120 7365 6372 6574  This is a secret
00000010: 206d 6573 7361 6765 2e0a                  message..
```

Encryption with camellia command and decryption with openssl command (CFB mode, the plaintext output has redundant trailing bytes):
```sh
$ echo "This is a secret message." | camellia --iv 0123456789abcdef --mode cfb -S 128 -R 0123456789abcdef | openssl enc -d -camellia-128-cfb -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | xxd
00000000: 5468 6973 2069 7320 6120 7365 6372 6574  This is a secret
00000010: 206d 6573 7361 6765 2e0a 0606 0606 0606   message........
$
```

Encryption with camellia command and decryption with openssl command (OFB mode, the plaintext output has redundant trailing bytes):
```sh
$ echo "This is a secret message." | camellia --iv 0123456789abcdef --mode ofb -R 0123456789abcdef | openssl enc -d -camellia-128-ofb -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | xxd
00000000: 5468 6973 2069 7320 6120 7365 6372 6574  This is a secret
00000010: 206d 6573 7361 6765 2e0a 0606 0606 0606   message........
$
```

Encryption with camellia command and decryption with openssl command (CTR mode, the plaintext output has redundant trailing bytes):
```sh
$ echo "This is a secret message." | camellia --iv 0123456789abcdef --mode ctr -R 0123456789abcdef | openssl enc -d -camellia-128-ctr -K 30313233343536373839616263646566 -iv 30313233343536373839616263646566 | xxd
00000000: 5468 6973 2069 7320 6120 7365 6372 6574  This is a secret
00000010: 206d 6573 7361 6765 2e0a 0606 0606 0606   message........
$
```

### Environment
- Operating System  
  Ubuntu 20.04.6 LTS (WSL2)
- Python Version
  3.8.10
- Refinery Version
  0.5.8

**Comment by huettenhain** [comment]
Hey @nmantani and thank you for the very detailed bug report. It is a more than reasonable goal to be compatible with OpenSSL, I will get on this right away.

**Comment by huettenhain** [comment]
I added tests in d95e57d6c573b0a76c389e55be32c82d5eda3cea and 3aafbef1a5cec2062e82cc75248822003015b892 /  8844dd32d256cae5e94bca8e37695dc7c96580c7 should fix it. The test suite is currently running, and I will release a new version if all checks go green.

**Comment by huettenhain** [comment]
This is fixed and was released to PyPi as `binary-refinery=0.5.9`, so I recommend that you update to that version. 

In summary: As you correctly pointed out, insisting on block alignment is really only necessary for ECB and CBC (and PCBC) mode, so refinery will now use no padding by default for any other mode. This was simply an oversight on my part.

As far as I can tell this should fix your problem. Please re-open this issue if you encounter any other related problems.

Any many thanks for the bug report!

**Comment by nmantani** [comment]
Thank you so much!
